lightweight-term-count-update.php

```diff
@@ -12,79 +12,219 @@
  * @package 		Lightweight_Term_Count_Update
  */
 
+/**
+ * Main plugin class.
+ */
 class LTCU_Plugin {
 
-	// post stati which should be counted in term post counting.  Normally at least publish
-	public $counted_stati = array( 'publish' );		
+	/**
+	 * Post statuses which should be counted in term post counting. By default
+	 * this is [ 'publish' ], but it can be altered via the
+	 * `ltcu_counted_statuses` filter.
+	 *
+	 * @var array
+	 */
+	public $counted_statuses = array( 'publish' );
+
+	/**
+	 * Store the terms that have been incremented or decremented to avoid
+	 * duplicated efforts.
+	 *
+	 * @var array
+	 */
+	public $counted_terms = array();
+
+	/**
+	 * Store the singleton instance.
+	 *
+	 * @var My_Singleton
+	 */
+	private static $instance;
+
+	/**
+	 * Build the object.
+	 *
+	 * @access private
+	 */
+	private function __construct() {
+		/* Don't do anything, needs to be initialized via instance() method */
+	}
+
+	/**
+	 * Get the Singleton instance.
 
-	public function __construct() {
-		add_action( 'init', array( $this, 'init' ), 10, 0 );
+	 * @return My_Singleton
+	 */
+	public static function instance() {
+		if ( ! isset( self::$instance ) ) {
+			self::$instance = new static;
+			self::$instance->setup();
+		}
+		return self::$instance;
 	}
 
-	public function init() {
+	/**
+	 * Setup the singleton.
+	 */
+	public function setup() {
+		// Prevent core from counting terms.
+		wp_defer_term_counting( true );
```

Comment:
I like the idea and it seems to be working fine. I'm just not sure whether we should not check whether the counting is still deferred right before running our SQL queries. While checking whether `wp_defer_term_counting()` is still returning `true` might be a quick solution, it does not seem to be bulletproof.

We could perhaps hook the `edited_term_taxonomy` or `edit_term_taxonomy` actions and check whether they were called for the term and taxonomy we are currently updating before running our updates.

I don't think it's necessary for a v1 of the plugin, since it should be really easy to fix the wrong counts (basically just disable the plugin for a while and everything should get in sync sooner or later).